### PR TITLE
Add placeholder and label customization

### PR DESCRIPTION
**Pull request by adnanjelic**
# Why
In this PR we have added customization options for the placeholders, labels and the input text field (look and feel).

The PR is pretty big, but the improvements are significant for future ease of maintenance.

Some parts of the changes are just extracting code into separate files/classes so that takes a significant part of the total LOC. 

For easier review, I have commented on the code that was just extracted and does not need any additional review.

# How
Many improvements have been made, so here is a bullet list:

1. Created 2 new constructors (`InlinePaymentCard` and `RowsPaymentCard`) to separate layouts and make it easier for the users to use default and row implementations (these return the new `PaymentCardData` introduced in the previous PR). These new constructors do not accept the layout and the implementation is hidden from the users as the component already provides the expected layout. This also makes the specific default inline and row implementations scalable. In the future they can expand independently without bloating the API with unnecessary parameters that are not needed for both layouts. It also follows the Compose guidelines of creating specific constructors for different implementations of the underlying specific components.

2. The previously created `PaymentCardComponent` has been renamed to simple `PaymentCard` as it should be the default constructor for the custom layouts for the users. If you feel that this needs to be changed I am open to suggestions.

3. Added new classes and builders to provide defaults for placeholders and labels

4. User custom placeholder texts can now be provided to all component implementations (we provide the defaults that can be overridden)

5. Added new input components to the interface and implementation that accept label and placeholder Composables so the users have the ability to add any kind of Composables and customize them as much as they need

6. Added `TextFieldColors` parameter and support to the new input components so the user can customize the input look and feel as they want

7. Created a new `CustomTextField` that works together with the above-mentioned inputs with Composables.

8. Added many samples of the components and their usage, both inside the lib module (as Previews) and in the sample app

9. We also took the opportunity to clean up a little bit the code:
    - Better code/class structuring
    - Got rid of "god" classes that had mixed-up code by extracting parts of the code to other SRP classes/files smaller for better maintainability and reusability

## Final notes:

- The changes are done so the existing API users are not affected. 
- No current API name or signature is changed
- If we needed to add something to the existing constructors, we overloaded them not to change any method/constructor signature. 
- The users are not forced and can switch to use the new components whenever they feel to (to use the Composables instead of simple configurations for example.)

### Samples
| Sample app update | Samples in the lib | Samples in the lib |
| --------------------- | --------------------| ---------------------|
| image | image | image |

### CODE REVIEW

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/InlinePaymentCardInput.kt [hunk 1]
```diff
@@ -12,10 +12,21 @@ import androidx.compose.runtime.Composable
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.unit.dp
+import com.evervault.sdk.input.model.placeholder.PlaceholderDefaults
+import com.evervault.sdk.input.model.placeholder.PlaceholderTexts
+
+@Composable
+fun inlinePaymentCardInputLayout(): @Composable() (PaymentCardInputScope.(Modifier) -> Unit) =
+    { inline(modifier = it) }
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/InlinePaymentCardInput.kt:20]: Moved to the top as it is the entry point (did not make it internal as it was exposed before)

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/InlinePaymentCardInput.kt [hunk 2]
```diff
@@ -12,10 +12,21 @@ import androidx.compose.runtime.Composable
...
+
+@Composable
+fun inlinePaymentCardInputLayout(placeholderTexts: PlaceholderTexts): @Composable() (PaymentCardInputScope.(Modifier) -> Unit) =
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/InlinePaymentCardInput.kt:23]: a new method for providing `placeholderTexts` functionality to existing API users

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInput.kt [hunk 3]
```diff
@@ -206,221 +185,3 @@ fun PaymentCardInput(
     )
 }
 
-interface PaymentCardInputScope {
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInput.kt:209]: Moved outside

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
+package com.evervault.sdk.input.ui
+
+import androidx.compose.material3.TextFieldColors
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.text.TextStyle
+
+interface PaymentCardInputScope {
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt:8]: Moved out from `PaymentCardInput` and added new composable methods commented below

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    fun CardNumberField(modifier: Modifier)
+
+    @Composable
+    fun CardNumberField(modifier: Modifier, options: TextFieldOptions)
+
+    @Composable
+    fun CardNumberField(modifier: Modifier, options: TextFieldOptions, placeholder: String)
+
+    @Composable
+    fun CardNumberField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt:33]: New input that accepts Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    fun ExpiryField(modifier: Modifier)
+
+    @Composable
+    fun ExpiryField(modifier: Modifier, options: TextFieldOptions)
+
+    @Composable
+    fun ExpiryField(modifier: Modifier, options: TextFieldOptions, placeholder: String)
+
+    @Composable
+    fun ExpiryField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt:54]: New input that accepts Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    fun CVCField(modifier: Modifier)
+
+    @Composable
+    fun CVCField(modifier: Modifier, options: TextFieldOptions)
+
+    @Composable
+    fun CVCField(modifier: Modifier, options: TextFieldOptions, placeholder: String)
+
+    @Composable
+    fun CVCField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScope.kt:75]: New input that accepts Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+import androidx.compose.ui.text.input.ImeAction
+import androidx.compose.ui.text.input.KeyboardCapitalization
+import androidx.compose.ui.text.input.KeyboardType
+import androidx.compose.ui.text.input.TextFieldValue
+import androidx.compose.ui.text.input.VisualTransformation
+import androidx.compose.ui.unit.dp
+import com.evervault.sdk.input.model.placeholder.PlaceholderTextsDefaults
+import com.evervault.sdk.input.ui.component.CustomTextField
+
+internal class PaymentCardInputScopeImpl(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt:32]: Moved out from `PaymentCardInput` and added new composable implementation methods commented below

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            state = creditCardNumber,
+            placeholder = placeholder,
+            modifier = modifier.focusRequester(creditCardRequester),
+            options = options,
+            onNext = { expiryRequester.requestFocus() },
+        )
+    }
+
+    @Composable
+    override fun CardNumberField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt:96]: New input that accepts Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+            state = expiryDate,
+            placeholder = placeholder,
+            modifier = modifier.focusRequester(expiryRequester),
+            options = options,
+            onNext = { cvcRequester.requestFocus() },
+        )
+    }
+
+    @Composable
+    override fun ExpiryField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt:149]: New input that accepts Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        CustomTextField(
+            state = cvc,
+            placeholder = placeholder,
+            modifier = modifier.focusRequester(cvcRequester),
+            options = options,
+        )
+    }
+
+    @Composable
+    override fun CVCField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt:203]: New input that accepts Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+            label = label,
+            placeholder = placeholder,
+            textStyle = textStyle,
+            textFieldColors = textFieldColors
+        )
+    }
+
+    @OptIn(ExperimentalMaterial3Api::class)
+    @Composable
+    private fun CustomTextField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/PaymentCardInputScopeImpl.kt:222]: Nothing new, just extracted from `PaymentCardInput`

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/RowsPaymentCardInput.kt [hunk 13]
```diff
@@ -3,26 +3,34 @@ package com.evervault.sdk.input.ui
...
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
-import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.text.style.TextAlign
 import androidx.compose.ui.unit.dp
+import com.evervault.sdk.input.model.placeholder.PlaceholderDefaults
+import com.evervault.sdk.input.model.placeholder.PlaceholderTexts
+
+@Composable
+fun rowsPaymentCardInputLayout(): @Composable() (PaymentCardInputScope.(Modifier) -> Unit) =
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/RowsPaymentCardInput.kt:23]: Moved to the top as it is the entry point (did not make it internal as it was exposed before)

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/RowsPaymentCardInput.kt [hunk 14]
```diff
@@ -3,26 +3,34 @@ package com.evervault.sdk.input.ui
...
+    { rows(modifier = it) }
+
+@Composable
+fun rowsPaymentCardInputLayout(placeholderTexts: PlaceholderTexts): @Composable() (PaymentCardInputScope.(Modifier) -> Unit) =
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/RowsPaymentCardInput.kt:27]: a new method for providing `placeholderTexts` functionality to existing API users

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/component/CustomTextField.kt [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+import androidx.compose.ui.text.input.TextFieldValue
+import androidx.compose.ui.text.input.VisualTransformation
+import androidx.compose.ui.unit.dp
+
+/**
+ * Input field that accepts composables for the placeholder and labels, so the user can modify and provide them as they need.
+ */
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+internal fun CustomTextField(
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/component/CustomTextField.kt:29]: new input field that works with Composables

File: evervault-inputs/src/main/java/com/evervault/sdk/input/ui/sample/PaymentCardSamples.kt [hunk 16]
```diff
@@ -0,0 +1,231 @@
+package com.evervault.sdk.input.ui.sample
```
Comment (adnanjelic) [inline evervault-inputs/src/main/java/com/evervault/sdk/input/ui/sample/PaymentCardSamples.kt:1]: Renamed the `PaymentCardInputComponentSamples`, modified and added new samples

File: sampleapplication/src/main/java/com/evervault/sampleapplication/sample/PaymentCardSamples.kt [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+/**
+ * This file is a partial copy of the PaymentCardWithCustomizedLayoutSample and private methods
+ * from com.evervault.sdk.input.ui.sample.PaymentCardSamples
+ *
+ * In the future we should create a shared module so some of the components used for previews can be
+ * shared without exposing them to the clients with public visibilities
+ */
+
+@Composable
+internal fun PaymentCardInputScope.PaymentCardCustomLayoutWithNewComponents(modifier: Modifier) {
```
Comment (adnanjelic) [inline sampleapplication/src/main/java/com/evervault/sampleapplication/sample/PaymentCardSamples.kt:33]: Needed to create this one not to pollute the `NavigationGraph`. Temporarily until we create a shared module where we can share components between the libs and the sample app

File: sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/InfoBlock.kt [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Column
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.unit.dp
+
+@Composable
+internal fun InfoBlock(label: String, value: String) {
```
Comment (adnanjelic) [inline sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/InfoBlock.kt:11]: Just extracted from `CreditCardInputView`

File: sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/StubbedCustomizations.kt [hunk 19]
```diff
@@ -0,0 +1,25 @@
+package com.evervault.sampleapplication.ui.views.component
+
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.graphics.Color
+import com.evervault.sdk.input.model.placeholder.PlaceholderDefaults
+
+@Composable
+internal fun CustomTheme(content: @Composable () -> Unit) {
```
Comment (adnanjelic) [inline sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/StubbedCustomizations.kt:9]: Extracted from `CreditCardInputView` to make it reusable

File: sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/StubbedCustomizations.kt [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+            primary = Color.Black,
+            secondary = Color.White,
+            primaryContainer = Color.LightGray,
+        ),
+        content = content
+    )
+}
+
+@Composable
+internal fun customPlaceholderTexts() = PlaceholderDefaults.texts(
```
Comment (adnanjelic) [inline sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/StubbedCustomizations.kt:21]: This is just to provide different placeholder texts than defaults to see that the user can customize them

File: sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/SupportingText.kt [hunk 21]
```diff
@@ -0,0 +1,13 @@
+package com.evervault.sampleapplication.ui.views.component
+
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.text.font.FontStyle
+
+@Composable
+internal fun SupportingText(text: String) {
```
Comment (adnanjelic) [inline sampleapplication/src/main/java/com/evervault/sampleapplication/ui/views/component/SupportingText.kt:8]: Simple component to wrap the supporting text (below the title in the `ListItem`) to make the `NavigationGraph` easier to read and maintain

**Review by boilsquid — APPROVED** [review]
Very nice!

**Pull request merged**